core/src/services/swift/core.rs

```diff
@@ -0,0 +1,237 @@
+    pub async fn swift_delete(&self, path: &str) -> Result<Response<IncomingAsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+
+        let url = format!(
+            "{}/v1/{}/{}/{}",
+            self.endpoint,
+            self.account,
+            self.container,
+            percent_encode_path(&p)
+        );
+
+        let mut req = Request::delete(&url);
+
+        req = req.header("X-Auth-Token", &self.token);
+
+        let request_body = &json!({
+            "path": percent_encode_path(&p),
+            "recursive": true,
+        });
+
+        let body = AsyncBody::Bytes(Bytes::from(request_body.to_string()));
+
+        let req = req.body(body).map_err(new_request_build_error)?;
+
+        self.client.send(req).await
+    }
+
+    pub async fn swift_list(&self, path: &str) -> Result<Response<IncomingAsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+
+        // The delimiter is used to disable recursive listing.
+        // Swift returns a 200 status code when there is no such pseudo directory in prefix.
+        let url = format!(
+            "{}/v1/{}/{}/?prefix={}&delimiter=/&format=json",
+            self.endpoint,
+            self.account,
+            self.container,
+            percent_encode_path(&p)
+        );
+
+        let mut req = Request::get(&url);
+
+        req = req.header("X-Auth-Token", &self.token);
+
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.client.send(req).await
+    }
+
+    pub async fn swift_create_object(
+        &self,
+        path: &str,
+        body: AsyncBody,
+    ) -> Result<Response<IncomingAsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+
+        let url = format!(
+            "{}/v1/{}/{}/{}",
+            self.endpoint,
+            self.account,
+            self.container,
+            percent_encode_path(&p)
+        );
+
+        let mut req = Request::put(&url);
+
+        req = req.header("X-Auth-Token", &self.token);
+
+        if p.ends_with('/') {
+            req = req.header("Content-Length", "0");
+        }
+
+        let req = req.body(body).map_err(new_request_build_error)?;
+
+        self.client.send(req).await
+    }
+
+    pub async fn swift_read(
+        &self,
+        path: &str,
+        range: BytesRange,
+    ) -> Result<Response<IncomingAsyncBody>> {
+        let p = build_abs_path(&self.root, path)
+            .trim_end_matches('/')
+            .to_string();
+
+        let url = format!(
+            "{}/v1/{}/{}/{}",
+            self.endpoint,
+            self.account,
+            self.container,
+            percent_encode_path(&p)
+        );
+
+        let mut req = Request::get(&url);
+
+        req = req.header("X-Auth-Token", &self.token);
+
+        let range = match (range.offset(), range.size()) {
+            (Some(offset), Some(size)) => {
+                format!("bytes={}-{}", offset, offset + size - 1)
```

Comment:
Please don't build range header by hand, use `BytesReange::to_header()` instead.

Reply:
Changed to `is_full()`